fix(flutter/a11y assessments): h1 missing a11y from each page on the web app

essjay05 opened the pull request:
Adds Semantic Widget to wrap the contents of each page's AppBar title attribute and adds header: true and headingLevel: 1 as Semantic attributes so the content of the title is compiled as an h1 tag to meet accessibility guidelines that each page must have a h1 tag present. Also updates the test cases for the home page and each use case page to check for its respective h1.

[Before Screenshot - Accessibility Assessments Home Page](https://screenshot.googleplex.com/4i9LuiGwvLnEcZ8)
[Before Screenshot - Checkbox List Title -- note: each use case page is missing an h1](https://screenshot.googleplex.com/3qQjfqvAMTehRsm)
[After Screenshot - Accessibility Assessments Home Page](https://screenshot.googleplex.com/APSJJXBmwNBP35m)
[After Screenshot - Checkbox List Title-- note: change is similar across each Accessibility use case page](https://screenshot.googleplex.com/6EGgZnTusEgeN5L)

Fixes b/338035526

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

essjay05 committed:
fix(flutter-a11y): each page is missing an h1 tag

Updates each page (home, and each use case page) to use the
Semantics widget and add the , and
 for each title in the AppBar to serve as
each page's appropriate h1.

Fixes b/338035526

essjay05 committed:
test(material/autocomplete): adding test to check for one h1 on autocomplete page

Updating auto_complete_test to see if it will find the one h1 semantics tag applied
to the auto_complete.dart use case page.

essjay05 committed:
test(flutter a11y_assessments demos): adding tests to check for h1 on demo pages

Updated tests for badge, checkboxListTile, and auto_complete to check for the
text string that I used in the Semantics widget with header and headingLevel
attributes as there doesn't seem to be a way to check for a header or headingLevel.

essjay05 committed:
test(flutter/a11y_assessments): implementing semantics h1 and tests to use case pages

Updates each use case page title in the AppBar so that it is using a Semantics widget
with header and headingLevel attributes to create an h1 title for each page. Also
updates the tests associated with each use case page to check for the existence of
an h1 on each page.

chunhtai commented on dev/a11y_assessments/lib/main.dart:
```diff
@@ -81,7 +81,13 @@ class HomePageState extends State<HomePage> {
   Widget build(BuildContext context) {
     return Scaffold(
-      appBar: AppBar(title: const Text('Accessibility Assessments')),
+      appBar: AppBar(
```
Instead of changing all example, should we directly bake headinglevel into Appbar widget?

essjay05 replied:
@chunhtai is [this change\(https://github.com/flutter/flutter/pull/152198/commits/3b9f9b260681ddd69740a651a789f7d6058889a3) sufficient for this?

essjay05 replied:
Was attempting to fix the a11y_assessments separately and then attempt to do the AppBar with `headingLevel: 1` baked in after? Open to whatever you think is best @chunhtai.

chunhtai replied:
we should do https://github.com/flutter/flutter/pull/152198/commits/3b9f9b260681ddd69740a651a789f7d6058889a3 instead

essjay05 replied:
Updated all use_case widgets to revert to previous and rely on app_bar default `headingLevel: 1`, all new tests still passed.

essjay05 committed:
fix(flutter/app_bar): by default set title header to headingLevel: 1 in app_bar

Updates material/app_bar when excludeHeaderSemantics is false, the
for app_bar, set the  by default to meet one h1 per page
accessibility guidelines.

essjay05 committed:
refactor(flutter/app_bar & flutter/nav_bar): updates to add headingLevel: 1 as header default

Updates material/app_bar and cupertino/nav_bar for when there is a Semantic tag associated
with the header & title to apply a default  to have a default h1 tag
applied to each page to meet accessibility guidelines. Adds comments to explain the default
reason.

essjay05 committed:
refactor(flutter/auto_complete): remove hard-coded headingLevel: 1

Updates previous fix to implement  into the use
case auto_complete appBar previous state now that packages/app_bar
by default applies  to the title value. Checked
and the auto_complete test still works. WIP: removing hard coded
 from all use_case widgets.

essjay05 committed:
refactor(flutter/a11y_assessments): refactor headingLevel: 1 to appBar

Update previous fix where it manually added  to each
use_case widget appBar, but refactored this into package/app_bar
instead to make it the default. Ran flutter tests on a11y_assessments
successfully.

chunhtai commented on packages/flutter/lib/src/material/app_bar.dart:
```diff
@@ -991,6 +991,9 @@ class _AppBarState extends State<AppBar> {
           header: true,
+          // set default headingLevel: 1 for one h1 element
+          // per page to meet accessibility guidelines
+          headingLevel: 1,
```
do we need this change to pass gar? I thought if we convert header to banner, It should have worked?

essjay05 replied:
I believe so, the [bug](b/338035526) I'm trying to fix references [GAR 1.6](go/GAR-web#overview-of-criterion-16) where it is required to have at least one h1 tag on each page. 

Let me know if this is a good way to enforce this rule.

essjay05 replied:
Attempted to make related engine changes for [issue](https://github.com/flutter/flutter/issues/152268) in [this PR](https://github.com/flutter/engine/pull/54224).

chunhtai replied:
This seems not addressed yet

chunhtai reviewed (commented):
waiting for review comment to be resolved

essjay05 committed:
fix(flutter/app_bar flutter/nav_bar): remove default headingLevel: 1

Updates previous fix to remove headingLevel:1 as the default headingLevel
in AppBar and NavBar.

essjay05 commented:
> waiting for review comment to be resolved

Just updated to remove default `headingLevel: 1` in this commit: https://github.com/flutter/flutter/pull/152198/commits/1dc9aba75a390675cd92a5258e5908c46dc2dd78.

essjay05 committed:
fix(flutter/a11y_assessments): implement semantics widget with headingLevel: 1

Updates previous fix which removes default  and implements it into
each use_cases' AppBar title.

essjay05 committed:
fix(flutter/a11y_assessments): linux unused import removal

Updates to remove unused import in auto_complete_test.dart to fix
failing Linux test.

chunhtai reviewed (approved):
LGTM

essjay05 committed:
fix(flutter/a11y_assessments): add heading to newly added component demos

Updates newly added component demos drawer and navigation drawer to use
semantics wrapper with headingLevel: 1 around the appBar title text.
Also added respective unit tests.

hannah-hyj reviewed (approved):
LGTM
